### PR TITLE
Squid template for FireSquid release

### PR DESCRIPTION
**Pull request by dzhelezov**
Includes:
- Batch processor
- Local setup for FireSquid archives
- Updated instructions for typegen

NB! Before merging, the images in `archive/docker-compose.yml` should be updated with the official images published by @tmcgroul

### CODE REVIEW

**Comment by RaekwonIII** [comment]
Checked changes and checked out the repo, run the code. Only a couple of questions:

- is there a reason for removing `make build` from Makefile?
- similarly, what is the reason for having generated `TimestampSetCall` and `SystemAccountStorage`? They are not used, it seems. Is it to have an example of typegen generating one-of-each?

Bonus question (realized this afterwards):

do we want the BatchProcessor to be the _de facto_ preferred way to develop with Subsquid? The template repository acts as "onboarding" for anyone interacting with Subsquid for the first time, but the Batch processing is a bit more difficult than the "normal" one.

**Comment by eldargab** [comment]
>do we want the BatchProcessor to be the de facto preferred way to develop with Subsquid?

Yes, I think we want this, because ultimately every mapping project becomes limited by speed.

**Comment by belopash** [comment]
I think we need two templates. One for SubstrateBatchProcessor and one for SubstrateProcessor. Sometimes it's much easier to use classic handlers.

**Comment by eldargab** [comment]
Not separate, but just 2 processor implementations present in the template, with batch processor being the default at which all commands are pointing.

**Comment by belopash** [comment]
It gives a gain in synchronization speed, but still is less convenient for simple projects. Therefore, in my opinion, it should be used case of significant performance problems, but not by default.

**Comment by RaekwonIII** [comment]
> Not separate, but just 2 processor implementations present in the template, with batch processor being the default at which all commands are pointing.

I am not sure what you mean with this one. How would that work? Would you instantiate a `SubstrateProcessor`, but not use it, not have `processor.run()`? I don't understand...

I would have agreed with Alex, my follow-up suggestion would have been to create a "Batch template" just as we have created an "EVM processor".

**Comment by dzhelezov** [comment]
I do believe that BatchProcessor should be the de facto standard. Similar to any vector-type libraries like Matlab, it takes some time to switch from for loops to vector-based commands, but the outcome is definitely worth the effort. So I think we should use BatchProcessor for the template, and describe the other options in docs.subsquid.io and drop a link in README

**Comment by eldargab** [comment]
>How would that work? 

The processor file is just an entry point, you can have as many as you want and launch any one at will.

>my follow-up suggestion would have been to create a "Batch template"

Hard no to that.

> it should be used in case of significant performance problems, but not by default.

The problem here, is that even tiny change in spec requirement or minor misunderstanding of how chain works can turn "insignificant" performance issues to show stopping ones, forcing the full rewrite.

If developers want to trade minor initial inconveniences for the risk of doing full rewrite, it is their choice, but, at least, they deserve to know what kind of speed difference batch processing has to offer!

**Comment by belopash** [comment]
Ok, two processors files with two launch commands sounds good to me. This allows developers to make their own decision about what to use.

**Comment by dzhelezov** [comment]
I strongly oppose to the two processor option as it will make it confusing on how to deploy it. The goal of this template is NOT to show all the possible options available -- we have docs for that. It is rather a minimal project an interested developer can run end-to-end and deploy, so offering any options to choose here is a bad UX.

**Comment by dzhelezov** [comment]
Regarding multiple templates -- we have a whole section "Tutorials" in docs, where we will show multiple projects with real value showcasing different options -- from "Hello World" to a complex one with custom resolves. But it will reside in a separate repo(s) with a dedicated README and instructions

**Comment by eldargab** [comment]
>I strongly oppose to the two processor option as it will make it confusing on how to deploy it.

I don't think that one additional file can cause any confusion. On the contrary, it can help the developer to make an informed decision about which option to choose. 

Separate repo with various examples will have a great value, but redirecting to another repo, 
forcing to go through annoying clone, build, up, run steps just to have an idea about speed differences? Sounds a bit much to me...

**Comment by dzhelezov** [comment]
No one will compare the speed at the first touch! A non-batched example will be referenced in docs, together with other tutorials.

So the user flow is:

- Go to template
- Run and play around 
- At this point the dev should be convinced by the performance of Subsquid and that it's worth diving deeper
- Go to docs and explore how to customize it to the use-case the dev has

Template is basically a landing page and we should make it simple, streamlined, yet convincing. Once the "selling" part is done, the rest will be looked up in the docs.

I see no value in adding more information to the template which may blur our key value proposition.

File: archive/docker-compose.yml [hunk 1]
```diff
@@ -1,6 +1,8 @@
-version: "3.4"
+version: "3"
```
Comment (eldargab) [inline archive/docker-compose.yml:1]: This field is now obsolete and is not semantic.

File: archive/docker-compose.yml [hunk 2]
```diff
@@ -1,6 +1,8 @@
...
 
 services:
+  # We recommend using a CockroachDB cluster for production deployments
```
Comment (eldargab) [inline archive/docker-compose.yml:4]: We have a little experience to recommend it and don't offer such option in SaaS.

File: archive/docker-compose.yml [hunk 3]
```diff
@@ -9,75 +11,56 @@ services:
...
-      - DEV_MODE=true
-      - DB_NAME=indexer
-      - DB_HOST=db
-      - DB_USER=postgres
-      - DB_PASS=postgres
-      - DB_PORT=5432
-      - HYDRA_INDEXER_STATUS_SERVICE=http://indexer-status-service:8081/status
+      DATABASE_TYPE: postgres # set to  `cockroach` for CockroachDb
+    command: [
+       "-e", "wss://kusama-rpc.polkadot.io",
```
Comment (eldargab) [inline archive/docker-compose.yml:25]: No need to duplicate endpoints anymore. See both links below.

* https://github.com/subsquid/squid/blob/829ee656033c0057c9967963c78b54bf11f0e3d8/substrate-ingest/src/main.ts#L26
* https://github.com/subsquid/squid/blob/829ee656033c0057c9967963c78b54bf11f0e3d8/substrate-ingest/src/main.ts#L80

File: archive/docker-compose.yml [hunk 4]
```diff
@@ -9,75 +11,56 @@ services:
...
+      - db
+    image: tmcgroul/archive-gateway:32
     environment:
-      REDIS_URI: redis://redis:6379/0
-      PORT: 8081
+      DATABASE_MAX_CONNECTIONS: 5
+      RUST_LOG: "actix_web=info,actix_server=info"
+    command: [
+       "--database-url", "postgres://postgres:postgres@db:5432/squid-archive",
+       # "--evm-support" # uncomment to additionally for a native support of Frontier EVM
```
Comment (eldargab) [inline archive/docker-compose.yml:46]: uncomment for chains with Frontier EVM pallet

File: src/processor.ts [hunk 5]
```diff
@@ -1,96 +1,116 @@
...
+import {Account, HistoricalBalance} from "./model"
+import {BalancesTransferEvent} from "./types/events"
+
+
+const processor = new SubstrateBatchProcessor()
+    .setBatchSize(500)
+    .setDataSource({
+        // For locally-run archives:
+        // archive: 'http://localhost:8888/graphql'
+        // Lookup archive by the network name in the Subsquid registry
```
Comment (eldargab) [inline src/processor.ts:14]:
```
// Lookup archive by the network name in the Subsquid registry.
//
// To connect to local archive launched by `archive/docker-compolse.yml`, 
// replace registry lookup with 'http://localhost:8888/graphql'.
```

File: README.md [hunk 6]
```diff
@@ -275,22 +267,20 @@ Type definitions (`typesBundle`) can be given to squid tools in two forms:
...
+In contrast, the squid framework works only with plain js primitives and objects.
+This allows to decrease coupling and is also simply dictated by the fact, that
 there is not enough information in substrate metadata to distinguish between 
 interesting cases.
 
-Account addresses is one example where such difference shows up.
-From substrate metadata (and squid framework) point of view account address is simply a fixed length
-sequence of bytes. On other hand, polkadot.js creates special wrapper for account addresses which 
-aware not only of address value, but also of its 
+Handing account addresses is one example where such a difference shows up.
```
Comment (eldargab) [inline README.md:281]: Handing or handling, this addition doesn't seem right.

**Comment by RaekwonIII** [comment]
I do think that both options in one repo would make it more confusing, that is why I asked. And the question about generated types that are not then used was along the same direction.

OK on Batch being the standard going forward if this is what we want to and think it's worth it.

And OK on showing other options in the tutorial as well. I will reserve some time after Polkadot Decoded to create those.

**Review by RaekwonIII — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -145,10 +148,10 @@ npx sqd db create
...
+Event, call and runtime storage data come to mapping handlers as raw untyped json. 
+While it is possible to work with raw untyped json data, it's extremely error-prone and the json structure may change over time due to runtime upgrades.
 
-Squid framework provides tools for generation of type-safe, spec version aware wrappers around events, calls and runtime storage items. Typegen generates type-safe classes in `types/events.ts`, `types/calls.ts` and `types/storage.ts` respectively, with constructors taking `XXXContext` interfaces as the only argument. All historical runtime upgrades are accounted out of the box. A typical usage is as follows (see `src/processor.ts`):
+Squid framework provides tools for the generation of type-safe, spec version aware wrappers around events, calls and runtime storage items. Typegen generates type-safe classes in `types/events.ts`, `types/calls.ts` and `types/storage.ts` respectively, with constructors taking `XXXContext` interfaces as the only argument. All historical runtime upgrades are accounted out of the box. Typical usage is as follows (see `src/processor.ts`):
```
Comment (eldargab) [inline README.md:154]: >with constructors taking `XXXContext` interfaces as the only argument

No need to say this.

File: .env [hunk 8]
```diff
@@ -1,4 +1,5 @@
 DB_NAME=squid
 DB_PASS=squid
 DB_PORT=23798
+PROCESSOR_PROMETHEUS_PORT=3000
```
Comment (eldargab) [inline .env:4]: No need to take such a common port. No need to take non-ephemeral port at all.

File: README.md [hunk 9]
```diff
@@ -110,7 +113,7 @@ Full description of `schema.graphql` dialect is available [here](https://docs.su
 ### 2. Generate TypeORM classes
 
 Mapping developers use TypeORM [EntityManager](https://typeorm.io/#/working-with-entity-manager)
-to interact with target database during data processing. All necessary entity classes are
+to interact with the target database during data processing. All necessary entity classes are
 generated by the squid framework from `schema.graphql`. This is done by running `npx sqd codegen`
```
Comment (eldargab) [inline README.md:117]: `npx squid-typeorm-codegen`

**Review by eldargab — APPROVED** [review]
Please, have a look at comments, but otherwise feel free to merge.
We'll need another pass on this anyway...

**Comment by dzhelezov** [comment]
All good, except for the fact that we need to stick to the old cli commands (eg 'npx sqd codegen') for now due to SaaS compatibility issues.

**Comment by eldargab** [comment]
We can continue to have old `cli` program around to have `sqd db migrate` available in production, but this doesn't prevent us from using new ones for everything else.

It would be very bad to introduce additional breaking transition.

**Pull request merged**